src/sage/rings/lazy_series.py

```
@@ -657,6 +657,22 @@ def shift(self, n):
             sage: f.shift(3)
             1/(5^t) + 2/6^t
 
+        .. WARNING::
+
+            When working with an inexact power series (e.g., defined
+            by a function) that has not computed its order and shifting
+            by too much can result in a Laurent series that does not
+            know it is a Laurent series::
```

There is a grammatical / phrasing issue here, which I cannot quite pin down ("When ... shifting by too much can result ...")

More importantly, however, we should make it very clear that `fs` in the example below is broken, and we would actually prefer if we could raise an error, but we can't without breaking implicit definitions:
sage: L.<z> = LazyPowerSeriesRing(QQ)
sage: f = L(lambda n: n)
sage: g = f.shift(-2); g
1/z + 2 + 3*z + 4*z^2 + 5*z^3 + 6*z^4 + O(z^5)
sage: g.inverse()*g
1/z - 1 + 2*z - 3*z^2 + 5*z^3 - 8*z^4 + O(z^5)

---

The sentence is slightly hard to parse, but it is correct. The key point is the "and" gives 2 necessary-but-not-sufficient conditions.

I would not say it is broken, nor would I say it should return an error. The behavior is well-defined, but the parent (and subsequently the element) just doesn't know enough to be fully correct. Strictly speaking, it is kind of a garbage-in-garbage-out scenario in my mind.

I am against options 1 and 2; while things would be more consistent (and hence, more manageable in some ways), it would also lead to very subtle surprises. 3 is certainly reasonable, although not without some subtleties. Alt 1 is somewhat reasonable, but it will put more burden on users with undefined series. Alt 3 is too cumbersome to allow fraction fields. Alt 2 is certainly a reasonable thing.

Actually, the more I think about it, the most consistent thing would be to just kill off any terms below the minimal valuation by default. Then taking a slight twist on Alt 2, we add a keyword that always returns an element in the fraction field.[^1] The current design is definitely an ease-of-implementation, but perhaps it was a mistake (likely mine). Changing it to this would make it the most consistent with almost no surprises. If we went this route, one remaining question is would we want to include a deprecation or not. 

[^1]: Strictly speaking, the Laurent field that contains the fraction field when working over a ring and not a field...

---

(I don't understand the bit about `g.inverse()*g != 1` not being a bug, but I think that's not important)

I am all for consistent behaviour, it appears to me that inconsistencies which are convenient eventually become a pain.

Since Stream_shift shares the cache with its input stream, we will need a condition in `Stream_shift.__getitem__`.  I hope that this doesn't come with a performance penalty.

Also, `Stream_shift.__hash__` has to be adapted: `f.shift(-2)` and `f.shift(-1)` would then be different for `f = L(lambda n: 0 if n else 1)`.

---

I reworked the shifting and the error message. There is a slight inconsistency between the known and not-known to be constant series, but this would be very complicated to make consistent I think. It is also a fairly fringe thing IMO.

There's a bit of a choice between inconsistent (but documented) behavior and simplicity of code in this case. It isn't really a pain to work with as I doubt anyone has encountered this situation without it being an actual bug in their code. For that reason, I don't think we need a deprecation warning, but I did add a flag just in case there is some user that was using the old behavior (now they have to always have the result as a Laurent series though).

The hash isn't a problem. Non-equal things can have the same hash. (There is a _very_ slight performance penalty when put into hast table objects such as a `dict`, but hash collisions are a general fact of life.) There are many more things that are an actual problem, where they are equal as FPS but have different hashes due to different constructions. However, solving that is likely a near-impossible task and definitely outside of the scope of this ticket.

---

Concerning the hash, I was writing about equal things that would have different hash.

For `f = L(lambda n: 0 if n else 1)`, the series `f.shift(-2)` and `f.shift(-1)` are equal if we do not push them into the fraction field, but will have the same hash.

---

Yes, which is why I am a bit confused, equal implies the same hash is what we want.

---

They would have differen hash, no?
```
        return hash((type(self), self._series))
```
Ah, no, they would actually have the same hash.  We should maybe make a note in `Stream_shift` that `__hash__` must not include the shift.

---

I don't like the following:
```
sage: L.<x> = LazyPowerSeriesRing(QQ)
sage: f1 = 1/(1-x)
sage: f2 = L(lambda n: 1)
sage: f1.shift(-1)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In [17], line 1
----> 1 f1.shift(-Integer(1))

File ~/sage-trac/src/sage/rings/lazy_series.py:762, in LazyModuleElement.shift(self, n, use_fraction_field)
    760         if not init_coeff and not self._coeff_stream._constant:
    761             return P.zero()
--> 762     coeff_stream = Stream_exact(init_coeff,
    763                                 constant=self._coeff_stream._constant,
    764                                 order=valuation, degree=degree)
    765 else:
    766     coeff_stream = Stream_shift(self._coeff_stream, n)

File ~/sage-trac/src/sage/data_structures/stream.py:649, in Stream_exact.__init__(self, initial_coefficients, constant, degree, order)
    647 else:
    648     self._degree = degree
--> 649 assert order + len(initial_coefficients) <= self._degree
    651 # we remove leading and trailing zeros from
    652 # initial_coefficients
    653 
   (...)
    656 # different from constant, because __eq__ below would become
    657 # complicated otherwise
    658 for i, v in enumerate(initial_coefficients):

AssertionError: 
sage: f2.shift(-1)
1 + x + x^2 + x^3 + x^4 + x^5 + x^6 + O(x^7)
```

---

And I really do not like
```
sage: g = f2.shift(-1)
sage: g
1 + x + x^2 + x^3 + x^4 + x^5 + x^6 + O(x^7)
sage: g[-1]
1
```

---

Yes, that is really bad. It also exposes a number of other issues with overshifting that I don't have an easy or even moderate solution for right now. Creating a truncated shift stream is not easy if we want to share the cache without explicitly testing for the existence of a cache. The old behavior is starting to look attractive, albeit at the expense of a badly behaving answer if the user overshifts and doesn't use the optional argument. Or we otherwise just don't share the cache when we (might) truncate with perhaps some Python magic to convert to a usual stream if we eventually find that it is okay.